src/proc/mod.rs

```diff
@@ -212,7 +213,131 @@ impl crate::Binding {
     pub fn to_built_in(&self) -> Option<crate::BuiltIn> {
         match *self {
             Self::BuiltIn(bi) => Some(bi),
-            Self::Location(..) => None,
+            Self::Location { .. } => None,
+        }
+    }
+}
+
+impl super::Module {
+    /// Apply the usual default interpolation for vertex shader outputs and fragment shader inputs.
+    ///
+    /// For every `Binding` that is a vertex shader output or a fragment shader
+    /// input, and that has an `interpolation` or `sampling` of `None`, assign a
+    /// default interpolation and sampling as follows:
+    ///
+    /// - If the `Binding`'s type contains only 32-bit floating-point values,
+    ///   vectors of such, or structs of either of the above, default its
+    ///   interpolation to `Perspective` and its sampling to `Center`.
+    ///
+    /// - If the `Binding`'s type contains any integer values, mark its
+    ///   interpolation as `Flat`.
+    ///
+    /// This function is a utility front ends may use to satisfy the Naga IR's
+    /// requirement that all I/O [`Binding`s] from the vertex shader to the
+    /// fragment shader must have non-`None` `interpolation` values. This
+    /// requirement is meant to ensure that input languages' policies have been
+    /// applied appropriately.
+    ///
+    /// All the shader languages Naga supports have similar rules:
+    /// perspective-correct, center-sampled interpolation is the default for any
+    /// binding that can vary, and everything else either defaults to flat, or
+    /// requires an explicit flat qualifier/attribute/what-have-you.
+    ///
+    /// [`Binding`s]: super::Binding
+
+    pub fn apply_common_default_interpolation(&mut self) {
+        use super::{Binding, ScalarKind, Type, TypeInner};
+
+        fn type_is_variable(ty: Handle<Type>, arena: &Arena<Type>) -> bool {
```

Comment:
I don't think we actually need to detect this. All the varyings we care about are strictly scalars and vectors. So we'd just need to check the I/O structs, reach the leaf nodes in them (i.e. non-structs), and place the defaults according to the `ScalarKind`.

Reply:
Ah, okay. Clearly it's only the leaves that can have meaningful bindings.